### PR TITLE
Complete code refactor

### PR DESCRIPTION
**Pull request by simonepri**
## Changes
Sorry for this but I've almost rewrote the entire package. 🙈❤️

- Now windows time readings are now correct (time is 100-ns units not in ms)
- The history for wmic readings now is a LRU cache that auto-expire after 1 min of inactivity to avoid memory leaks
- CPU usage on windows should now be more accurate. If there isn't any history for a given pid, then the measurement is took twice.
- The `start` property of the statistics object has been removed. Actually it was completely wrong.  (BREAKING CHANGE)
- The `time` property has been renamed to `ctime` that means cpu time.  (BREAKING CHANGE)
- The `elapsed` has been added. It's the time in ms since the process has been started.  (BREAKING CHANGE)
- The `timestamp` property has been added. It's the time at which we took the measure.  (BREAKING CHANGE)
- The `unmonitor` and `_history` functions exported from the `index.js` has been complitely removed. Now we export just the stat funtion (BREAKING CHANGE)
- Tests are re-implemented and improved using `ava`
- Coverage report with `nyc` added
- Over-simplified the `readme.md` (We may want to re-add some documentation on how it works)
- Update `.gitignore` file
- Cleaned `package.json` file
- `procfile` implementation has been completely removed (We may consider to re-introduce this in the future if we will need it, but now is useless)
- The `ppid` property has been added.

I can understand that maybe is too much but I need this package to work.
If you are unhappy with the changes let me know! Thanks!

You can walk though the code [here](https://github.com/simonepri/pidusage/tree/refactor)

### CODE REVIEW

**Comment by simonepri** [comment]
@soyuka the PR should be ready. Tests are now solid and passing.
Maybe we should consider to add tests using Docker to test all the other non tested platforms.

**Comment by simonepri** [comment]
I've done some performance related improvements.
Let me know your opinion.

File: lib/wmic.js [hunk 1]
```diff
@@ -1,106 +1,118 @@
...
-      return done(new Error(stderr), null)
-    }
+    // Example of stdout on Windows 10
+    // CreationDate: is in the format yyyymmddHHMMSS.mmmmmmsUUU
+    // KernelModeTime: is in units of 100 ns
+    // UserModeTime: is in units of 100 ns
+    // WorkingSetSize: is in bytes
+    //
+    // CreationDate               KernelModeTime  ParentProcessId  ProcessId  UserModeTime  WorkingSetSize
+    // 20150329221650.080654+060  153750000       0                777        8556250000    110821376
```
Comment (soyuka) [inline lib/wmic.js]: Can you source your informations (where does format and unit of cpu times come from)?

File: lib/ps.js [hunk 2]
```diff
@@ -1,131 +1,103 @@
...
-        start: new Date(date - time * 1000)
+      var pid = parseInt(line[1], 10)
+      var ppid = parseInt(line[2], 10)
+      var cpu = parseFloat(line[3].replace(',', '.'), 10)
+      var memory = parseInt(line[4], 10) * 1024
+      var etime = parseTime(line[0])
+      var ctime = parseTime(line[5], true)
+
+      statistics[pid] = {
+        cpu: Math.min(Math.round(cpu * 1000) / 1000, 100.0),
```
Comment (soyuka) [inline lib/ps.js]: you said in the readme that it could be over 100? why don't just give back the results?
Also what's the point of the Math.round?

File: lib/ps.js [hunk 3]
```diff
@@ -1,131 +1,103 @@
...
 
     stdout = stdout.split(os.EOL)
 
     var statistics = {}
     for (var i = 1; i < stdout.length; i++) {
-      stdout[i] = stdout[i].trim()
+      var line = stdout[i].trim().replace(/\s\s+/g, ' ').split(' ')
 
-      if (!stdout[i]) {
+      if (!line || line.length !== 6) {
```
Comment (soyuka) [inline lib/ps.js]: If I understand this, you'll skip th eline if it hasn't all the data? How can this happend?

File: lib/ps.js [hunk 4]
```diff
@@ -1,131 +1,103 @@
...
-  * @param   {Object}           options
-  * @return  {Function}         callback(err, stat)
-  * on os x skip headers with pcpu=,rss=
-  * on linux it could be --no-header
-  * on solaris 11 can't figure out a way to do this properly so...
-  */
-function ps (pid, options, done) {
-  if (!Array.isArray(pid)) {
-    pid = [pid]
+function parseTime (timestr, fraction) {
```
Comment (soyuka) [inline lib/ps.js]: Please take a look at my algorithm for that it's more concise, just add the fraction to it and you're good to go. Also it doesn't use any regex (should be faster):

```
/**
 * Get ps etime in seconds
 * POSIX format should be: [[dd-]hh:]mm:ss
 * on linux take "etimes" which is already the number of seconds since process started
 *
 * @param {String}
 * @return seconds
 */
function parseTime (time) {
  // Example output 7-22:43:36
  var tmp = time.split(':')

  if (~tmp[0].indexOf('-')) {
    var daysAndHours = tmp[0].split('-')
    tmp[0] = daysAndHours[1] // hours
    tmp.unshift(daysAndHours[0]) // days
  }

  var l = tmp.length
  var k = Math.pow(60, l - 1)
  time = 0

  for (var i = 0; i < l; i++) {
    time += k * parseInt(tmp[i], 10)
    k /= 60
  }

  return time
}
```

File: lib/wmic.js [hunk 5]
```diff
@@ -1,106 +1,118 @@
...
+    var uptime = Math.floor(os.uptime())
 
-      return done(new Error(stderr), null)
-    }
+    // Example of stdout on Windows 10
+    // CreationDate: is in the format yyyymmddHHMMSS.mmmmmmsUUU
+    // KernelModeTime: is in units of 100 ns
+    // UserModeTime: is in units of 100 ns
+    // WorkingSetSize: is in bytes
+    //
```
Comment (soyuka) [inline lib/wmic.js]: Please add the information that results are coming back in their keys alphabetical order (useful if we add one to the future even though you did good at settings them at that order already above).

File: lib/wmic.js [hunk 6]
```diff
@@ -1,106 +1,118 @@
...
+      // process usage since last call
+      var total = (kerneltime + usertime - hst.ctime)
+      // time elapsed between calls
+      var seconds = uptime - hst.uptime
+      var cpu = seconds > 0 ? (total / (seconds * 1000)) * 100 : 0
+
+      history.set(pid, {ctime: usertime + kerneltime, uptime: uptime})
+
+      statistics[pid] = {
+        cpu: Math.min(Math.round(cpu * 1000) / 1000, 100.0),
```
Comment (soyuka) [inline lib/wmic.js]: same why the round thing? I tend to agree with the `Math.min` though or it could confuse end users.

File: lib/wmic.js [hunk 7]
```diff
@@ -1,106 +1,118 @@
...
+        ppid: ppid,
+        pid: pid,
+        ctime: usertime + kerneltime,
+        elapsed: date - creation.getTime(),
+        timestamp: date
+      }
     }
 
-    done(null, results)
+    if (again) {
```
Comment (soyuka) [inline lib/wmic.js]: This I kinda disagree because it makes the first call really slow no? I agree that the data will be accurate only from the second call on windows but we should just state so in the README.

File: package.json [hunk 8]
```diff
@@ -1,28 +1,41 @@
...
+  },
+  "scripts": {
+    "lint": "standard index.js lib/*.js test/*.js",
+    "test": "npm run lint&& nyc ava -m \"!*benchmark*\"",
+    "bench": "ava -m \"*benchmark*\""
+  },
   "dependencies": {},
   "devDependencies": {
-    "faucet": "0.0.1",
+    "ava": "*",
```
Comment (soyuka) [inline package.json]: Fancy hipster test framework with 81 dependencies... 
Tape: not fancy, just works 13 dependencies...

Also why no version constraint?

File: package.json [hunk 9]
```diff
@@ -1,28 +1,41 @@
...
+    "mockdate": "^2.0.2",
     "mockery": "1.4.0",
-    "nanobench": "^2.1.1",
-    "ps-tree-ce": "^1.1.0",
+    "nyc": "^11.6.0",
```
Comment (soyuka) [inline package.json]: what's this for?

File: test/bench.js [hunk 10]
```diff
@@ -1,47 +1,80 @@
...
+    for (let i = 0; i < times; i++) {
+      await pify(m)(pids)
+    }
+    const time = end()
+    return Promise.resolve(time)
+  } catch (err) {
+    end()
+    return Promise.reject(err)
+  }
+}
```
Comment (soyuka) [inline test/bench.js]: can we avoid using promises? Especially in benchmarks :p.

File: test/bench.js [hunk 11]
```diff
@@ -1,47 +1,80 @@
...
-    cb(list.filter(function (v) {
-      return !isNaN(parseInt(v))
-    }))
+      child.stdout.on('data', function (childs) {
+        if (++count === pidno) resolve(childs)
+      }.bind(this, childs))
+      child.stderr.on('data', function (data) {
+        reject(data.toString())
+      })
+      child.on('error', reject)
```
Comment (soyuka) [inline test/bench.js]: why do we need to listen to stderr?

File: test/helpers/mocks.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+    if (error) {
+      ee.emit('error', error);
+    } else if (!stderr) {
+      ee.stdout.on('end', () => ee.emit('close', code, signal));
+    } else {
+      ee.stderr.on('end', () => ee.emit('close', code, signal));
+    }
+    return ee;
+  }
```
Comment (soyuka) [inline test/helpers/mocks.js]: cool that you kept my mock stuff. Then again, can we avoid:
- es6 `import` 
- arrow functions
- lint this code with standard

File: test/integration.js [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+  t.false(isNaN(result.memory), 'memory')
+  t.is(typeof result.ppid, 'number', 'ppid')
+  t.false(isNaN(result.ppid), 'ppid')
+  t.is(typeof result.pid, 'number', 'pid')
+  t.false(isNaN(result.pid), 'pid')
+  t.is(typeof result.elapsed, 'number', 'elapsed')
+  t.false(isNaN(result.elapsed), 'elapsed')
+  t.is(typeof result.timestamp, 'number', 'timestamp')
+  t.false(isNaN(result.timestamp), 'timestamp')
+})
```
Comment (soyuka) [inline test/integration.js]: Now you make me learn a new test framework before all I used was:

```
t.ok()
t.equa()
t.deepEqual()
```

Again what's the benefit of using promises, await all this? Ok it's fancy but I wanted to keep this stuff workable on node 4 with no babel or fancy stuff...

File: test/integration.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+  await t.notThrows(pify(m)([65535, process.pid]))
+})
+
+test('should throw an error if the pid does not exists', async t => {
+  await t.throws(pify(m)([65535]))
+})
+
+test('should throw an error if the pid is too large', async t => {
+  await t.throws(pify(m)(99999999))
+})
```
Comment (soyuka) [inline test/integration.js]: does this test pass on windows?!

File: test/integration.js [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+  t.is(err.message, 'One of the pids provided is invalid')
+  err = await t.throws(pify(m)(['invalid']))
+  t.is(err.message, 'One of the pids provided is invalid')
+  err = await t.throws(pify(m)(-1))
+  t.is(err.message, 'One of the pids provided is invalid')
+  err = await t.throws(pify(m)([-1]))
+  t.is(err.message, 'One of the pids provided is invalid')
+})
+
+test('should not throw an error if one of the pids does not exists', async t => {
```
Comment (soyuka) [inline test/integration.js]: Not sure that's the behavior we agreed on. IMO it should just return blank statistics

File: README.md [hunk 16]
```diff
@@ -1,114 +1,147 @@
...
+<!-- Links -->
+[new issue]: https://github.com/soyuka/pidusage/issues/new
+[license]: https://github.com/soyuka/pidusage/tree/master/license
+[contributors]: https://github.com/soyuka/pidusage/contributors
 
-## Licence
+[github:soyuka]: https://github.com/soyuka
+[twitter:s0yuka]: http://twitter.com/intent/user?screen_name=s0yuka
+[github:simonepri]: https://github.com/simonepri
+[twitter:simoneprimarosa]: http://twitter.com/intent/user?screen_name=simoneprimarosa
```
Comment (soyuka) [inline README.md]: Same here, sorry but noise not needed.

File: README.md [hunk 17]
```diff
@@ -1,114 +1,147 @@
...
+- [pidusage-tree][gh:pidusage-tree] -
+Compute a pidusage tree
 
-```
-Benching 246 process
-NANOBENCH version 2
-> node test/bench.js
+## Authors
+- **Antoine Bluchet** -  *Follow* me on *Github* ([:octocat:@soyuka][github:soyuka]) and on  *Twitter* ([🐦@s0yuka][twitter:s0yuka])
+- **Simone Primarosa** -  *Follow* me on *Github* ([:octocat:@simonepri][github:simonepri]) and on  *Twitter* ([🐦@simoneprimarosa][twitter:simoneprimarosa])
```
Comment (soyuka) [inline README.md]: I really don't agree to add things like this on the readme its point being to explain the code not to sell myself or any of the collaborators.

File: index.js [hunk 18]
```diff
@@ -1,35 +1,3 @@
...
-      delete history[i]
-    }
-    return
-  }
-
-  delete history[pid]
-}
-
-exports._history = history
+module.exports = require('./lib/stats')
```
Comment (soyuka) [inline index.js]: I like that but it looks like the README has been wrongly updated? There is no `get` export. Although I'd be fine with the following API:

```
require('pidusage')(pids, [opts,] callback)
```

**Comment by soyuka** [comment]
> I can understand that maybe is too much but I need this package to work.

Everything is fine, note that this package in v1 is working just fine and is required by a lot of modules :).

I like the work thank you for putting time into this! 
A few general remarks:

- didn't we agreed on an array?
- I think that we should make some errors graceful, for example if the pid has not been found

With these things in mind a given array should give out the same lengthed array with statistics:

```javascript
pidusage([1,2,3], function (err, data) {
  // data 
  [
    {pid: 1, memory: 0 ...}, // assuming this pid has not been found, return blank statistics
    ...
  ]
})
```

- I'm not fond of the wmic/history thing with the self-invalidation. This should be left to the developer and it's out of scope.

Last but not least, please don't make me learn a new test framework :D

File: .travis.yml [hunk 19]
```diff
@@ -11,3 +11,5 @@ os:
 cache:
   directories:
   - node_modules
+after_success:
+- npm run bench
```
[the remark this answers is not in the document]
Comment (simonepri) [reply .travis.yml]: Yes, its just free. Why not?
We can see the bench on various platform.

File: lib/wmic.js (hunk 1 above)
Comment (simonepri) [reply lib/wmic.js]: The link is the same:
https://msdn.microsoft.com/en-us/library/aa394372(v=vs.85).aspx

Anyway I've double checked them on my bootcamp partition.

Comment (simonepri) [reply lib/wmic.js]: The format of CreationDate has been taken from here:
https://superuser.com/a/937401/470946

File: lib/ps.js (hunk 2 above)
Comment (simonepri) [reply lib/ps.js]: Because we know that it cannot be greater than 100.

The point of Math.round is to round to 2 decimal places.
The percentage is just an approximation there's no need to have more than 2 digits of precision.

File: lib/ps.js (hunk 3 above)
Comment (simonepri) [reply lib/ps.js]: It cannot, but who knows 🙈

File: lib/wmic.js (hunk 7 above)
Comment (simonepri) [reply lib/wmic.js]: If we don't do this the readings are just wrong.
And yes only the first call takes longer.

> I agree that the data will be accurate only from the second call

That's not true. Actually in the previous version you use the history also if that history was really old (like 10 hours ago) instead now the previous values are used only if they are not older than a minute. 

This means that:
- if you call pidusage frenquently you get "fast" and "correct" readings.
- if you call pidusage sporadically you get "correct" readings.

File: package.json (hunk 8 above)
Comment (simonepri) [reply package.json]: The output provided by ava is just far superior to tape.
Just watch the output from the CI.

In my opinion It makes the tests development much easier.

Also you can use ES6 syntax in tests also while testing on node versions that does not support it.

File: package.json (hunk 9 above)
Comment (simonepri) [reply package.json]: nyc is for the coverage report:
image

File: test/bench.js (hunk 10 above)
Comment (simonepri) [reply test/bench.js]: The overhead is insignificant.
Without promises the code would be really too complicate.

File: test/bench.js (hunk 11 above)
Comment (simonepri) [reply test/bench.js]: Because something can happen while spawning the processes.
It's here mainly because if the code inside the code variable is wrong we can catch the errors returned by node.

File: test/helpers/mocks.js (hunk 12 above)
Comment (simonepri) [reply test/helpers/mocks.js]: Oh I forgot to lint this sorry. Generally I use xo codestyle.

File: test/integration.js (hunk 13 above)
Comment (simonepri) [reply test/integration.js]: Why to bother about compatibility if this is handled by ava internally?
Tests are meant to be easy to write and read.

File: test/integration.js (hunk 14 above)
Comment (simonepri) [reply test/integration.js]: Yes!

File: README.md (hunk 17 above)
Comment (simonepri) [reply README.md]: No preference over this. I just used my default template 👍

File: index.js (hunk 18 above)
Comment (simonepri) [reply index.js]: Nice catch tanks!

File: .npmrc [hunk 20]
```diff
@@ -0,0 +1 @@
+package-lock=false
```
[the remark this answers is not in the document]
Comment (simonepri) [reply .npmrc]: The package don't has any dependency. We don't need it for the devDep.

File: test/integration.js (hunk 15 above)
Comment (simonepri) [reply test/integration.js]: Yes It returns partial statistics. But it's important that it shouldn't throw an error.

File: lib/ps.js (hunk 4 above)
Comment (simonepri) [reply lib/ps.js]: Actually `pow` is one of the slowest operation you can use on a computer 😅just like `log`.
Anyway how do you want to handle the fraction?

File: README.md [hunk 21]
```diff
@@ -1,114 +1,147 @@
...
+  //   1234: {
+  //     cpu: 0.1,             // percentage
+  //     memory: 3846144,      // bytes
+  //     ppid: 727,            // PPID
+  //     pid: 1234,            // PID
+  //     ctime: 0,             // ms user + system time
+  //     elapsed: 20000,       // ms since the start of the process
+  //     timestamp: 864000000  // ms since epoch
+  //   }
+  // }
```
[the remark this answers is not in the document]
Comment (simonepri) [reply README.md]: Actually is a bit complicated than we thought.

When you loop over the results you don't know in which index you need to place the result in the output array.
In order to know that you need to build an hashmap and this would be too expensive and useful on our side.

Eg:
The user gives you: `pids = [123, 412, 124]`
You need to create:
`map = {'123':0, '412':1, '124':2}`

Then you can do
`output = []`
`output[map[pid]] = stats`

Since we need to create the map there's no reason to prefer the array solution that would also be slower for us.

**Comment by simonepri** [comment]
> this package in v1 is working just fine

I don't think so.
(Sorry to be rude)
I must agree that your work is the best so far on the npm, but also now after the refactoring is far to be perfect.

> and is required by a lot of modules :)

This is the reason why we need to have rock-solid tests for this fundamental module

File: .travis.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (soyuka) [reply .travis.yml]: Noone is ever going to look at that IMO. Furthermore they'll not help us prevent performances regression if we can't compare benchmarks no?

File: lib/ps.js (hunk 3 above)
Comment (soyuka) [reply lib/ps.js]: Then remove this part from the condition :laughing:

File: lib/wmic.js (hunk 1 above)
Comment (soyuka) [reply lib/wmic.js]: cool, could you add these links to the comments?

File: lib/ps.js (hunk 4 above)
Comment (soyuka) [reply lib/ps.js]: I'm not sure, I just benchmarked our two functions and mine looks faster:

```
# yours
ok ~314 μs (0 s + 313642 ns)

# mine
ok ~148 μs (0 s + 148158 ns)
```

Comment (soyuka) [reply lib/ps.js]: I've added fractional seconds:

```
function parseTimeDos (time) {
  // Example output 7-22:43:36
  var tmp = time.split(':')

  if (~tmp[0].indexOf('-')) {
    var daysAndHours = tmp[0].split('-')
    tmp[0] = daysAndHours[1] // hours
    tmp.unshift(daysAndHours[0]) // days
  }

  var l = tmp.length
  var k = Math.pow(60, l - 1)
  time = 0

  for (var i = 0; i < l; i++) {
    time += k * parseInt(tmp[i], 10)
    k /= 60
  }

  if (PLATFORM === 'darwin') {
    var frac = tmp[i - 1].split('.')
    time += parseInt(frac[1], 10) * 600
  }

  return time * 1000
}
```

```
# yours
ok ~310 μs (0 s + 310036 ns)

# mine
ok ~196 μs (0 s + 196210 ns)
```

Comment (simonepri) [reply lib/ps.js]: Cool! The power of benchmarking.
Put them in a loop and run them at least 1000 times if you can. I'm curious about the results.

File: .travis.yml (hunk 19 above)
[the remark this answers is not in the document]
Comment (soyuka) [reply .travis.yml]: Also it makes CI go slower. We should call benchs when we need them :).

File: lib/ps.js (hunk 4 above)
Comment (soyuka) [reply lib/ps.js]: sure men, test data: 
```
var data = ['1-02:03:04.12', '1-01:02:03.12', '13:27.12', '01:02:03.12', '0:00']
```

1000 times * 4:

```

# one
ok ~7.48 ms (0 s + 7484277 ns)

# two
ok ~6.72 ms (0 s + 6715629 ns)

```

There's always a significant enough gap.

There is the bench code:

```
var PLATFORM = 'darwin'

function parseTime (timestr, fraction) {
  var time = 0
  var tpart = timestr.replace(/-|\./g, ':').split(':')
  var i = tpart.length - 1
  if (i >= 0 && fraction && PLATFORM === 'darwin') { // Fractions of second
    time += parseInt(tpart[i--], 10) * 600
  }
  if (i >= 0) { // Seconds
    time += parseInt(tpart[i--], 10) * 1000
  }
  if (i >= 0) { // Minutes
    time += parseInt(tpart[i--], 10) * 60000
  }
  if (i >= 0) { // Hours
    time += parseInt(tpart[i--], 10) * 3600000
  }
  if (i >= 0) { // Days
    time += parseInt(tpart[i--], 10) * 86400000
  }
  return time
}


/**
 * Get ps etime in seconds
 * POSIX format should be: [[dd-]hh:]mm:ss
 * on linux take "etimes" which is already the number of seconds since process started
 *
 * @param {String}
 * @return seconds
 */
function parseTimeDos (time) {
  // Example output 7-22:43:36
  var tmp = time.split(':')

  if (~tmp[0].indexOf('-')) {
    var daysAndHours = tmp[0].split('-')
    tmp[0] = daysAndHours[1] // hours
    tmp.unshift(daysAndHours[0]) // days
  }

  var l = tmp.length
  var k = Math.pow(60, l - 1)
  time = 0

  for (var i = 0; i < l; i++) {
    time += k * parseInt(tmp[i], 10)
    k /= 60
  }

  if (PLATFORM === 'darwin') {
    var frac = tmp[i - 1].split('.')
    time += parseInt(frac[1], 10) * 600
  }

  return time * 1000
}

var bench = require('nanobench')

var data = ['1-02:03:04.12', '1-01:02:03.12', '13:27.12', '01:02:03.12', '0:00']

bench('one', function (b) {
  b.start()

  for (let i = 0; i < 1000; i++) {
    data.forEach(function (d) {
      parseTime(d)
    })
  }

  b.end()
})

bench('two', function (b) {
  b.start()

  for (let i = 0; i < 1000; i++) {
    data.forEach(function (d) {
      parseTimeDos(d)
    })
  }

  b.end()
})
```

I run in with `node bench.js`

**Comment by soyuka** [comment]
> The problem is the nature of pids to disappear on a system, throwing an error is not appropriate.
If the user didn't get a reading then it can understand that that pid is not present in the system.
Obvs we need to document that behaviour!

Mhh so, if I call pidusage with an array of pids and one of them isn't found. Will this give back partial results AND an error? 
I ask because I really need pidusage to return empty statistics if the process has not been found, not to send an error. 
Maybe we can pass an option to still get an error when the process does not exist though.

> What are the advantages of tap?
I believe that the tests would be more complicated to write.

Less noise, less fancy just good ol' non-breaking working for years stuff :). Sure thing let's keep this fancy ava thing...

Good work on the README it works like that for me!

So left to do:

1. see the parseTime stuff
2. Add the wmic link to the comment
3. Question about the statistics when pid has not been found

Thanks for your work, looks great so far!

File: README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (soyuka) [reply README.md]: Okay I agree with that, also makes things easier in the results parsing!

File: lib/ps.js (hunk 4 above)
Comment (soyuka) [reply lib/ps.js]: you really don't like my function do you :laughing:

File: .npmrc (hunk 20 above)
[the remark this answers is not in the document]
Comment (soyuka) [reply .npmrc]: It does speed up things for the CI when installing dev deps no?

File: lib/ps.js [hunk 22]
```diff
@@ -7,7 +7,7 @@ var PLATFORM = os.platform()
 
 function parseTime (timestr, fraction) {
   var time = 0
-  var tpart = timestr.replace(/-|\./g, ':').split(':')
+  var tpart = timestr.split(/-|:|\./)
```
Comment (soyuka) [inline lib/ps.js]: now yours faster :D

File: lib/ps.js (hunk 4 above)
Comment (simonepri) [reply lib/ps.js]: I just want to understand why is faster 😅
Can you re-run the bench
replacing
`var tpart = timestr.replace(/-|\./g, ':').split(':')`
with
`var tpart = timestr.split(/-|:|\./)`

Thanks
If it still faster then we will adopt it

File: lib/wmic.js (hunk 1 above)
Comment (simonepri) [reply lib/wmic.js]: Done

File: lib/wmic.js (hunk 5 above)
Comment (simonepri) [reply lib/wmic.js]: done

File: lib/ps.js (hunk 22 above)
Comment (simonepri) [reply lib/ps.js]: 🎊How much is improved?

File: index.js [hunk 23]
```diff
@@ -1,3 +1,30 @@
...
+ */
+function pidusage (pids, callback) {
+  if (typeof callback === 'function') {
+    stats(pids, callback)
+    return
+  }
+  return pify(stats, pids)
+}
+
+module.exports = pidusage
```
Comment (soyuka) [inline index.js]: I just did that as well haha. let me rebase... We could've done this in another patch

**Comment by simonepri** [comment]
> Good work on the README it works like that for me!

Should we need to document how it works under the hood?
Like in the previous version?

> I really need pidusage to return empty statistics if the process has not been found, not to send an error.

That's the current behaviour if you pass an array of pids.
Instead if you pass a single pid then the error 'No maching pid found' is given.

That's the default behaviour of ps and wmic also

File: index.js (hunk 23 above)
Comment (simonepri) [reply index.js]: Sorry! 🙈

**Comment by simonepri** [comment]
Also lets consider to enables services like `codecov` and `snyk` and add their badges to the readme

EG:
https://codecov.io/gh/simonepri/pidtree
https://snyk.io/test/github/simonepri/pidtree?targetFile=package.json

**Comment by simonepri** [comment]
@soyuka can you create a PR with the changes you want on my fork instead of working on your branch
In that way I can follow the changes.

Then we will squash commits while merging this.

**Comment by soyuka** [comment]
I squashed and merged in #51 !! Thank you very much for the work, I'll open maybe a couple more merge requests this week to keep improving while I'm working on https://github.com/Unitech/pm2/pull/3546 !

**Comment by simonepri** [comment]
@soyuka I cannot see the changes for the readme. I believe that something went wrong while merging

**Comment by soyuka** [comment]
Oh yes thanks, was a mess to rebase because you merged master back to your branch. Prefer rebase next time :). All good now I think!

**Comment by soyuka** [comment]
I activated https://codecov.io/gh/soyuka/pidusage not sure if I should configure something

**Comment by simonepri** [comment]
Yes you should add
`npx codecov`
inside `travis.yml` and `appveyor.yml`
Like this:
https://github.com/simonepri/pidtree/blob/master/.travis.yml
